crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2389,28 +2414,48 @@ impl<'db> TypeInferenceBuilder<'db> {
             )
         };
 
-        let unbound_ty = if may_be_unbound {
-            Some(self.lookup_name(name))
-        } else {
-            None
-        };
+        let mut definitions = definitions.peekable();
+        let has_definitions = definitions.peek().is_some();
 
-        let ty = bindings_ty(self.db, definitions, unbound_ty);
-        if ty.is_unbound() {
-            self.diagnostics.add(
-                name.into(),
-                "unresolved-reference",
-                format_args!("Name `{id}` used when not defined"),
-            );
-        } else if ty.may_be_unbound(self.db) {
-            self.diagnostics.add(
-                name.into(),
-                "possibly-unresolved-reference",
-                format_args!("Name `{id}` used when possibly not defined"),
-            );
+        let bindings_ty = bindings_ty(self.db, definitions);
+        if may_be_unbound {
+            match self.lookup_name(name) {
+                SymbolLookupResult::Type(ty, _) => match bindings_ty {
+                    SymbolLookupResult::Type(bindings_ty, _) => {
+                        UnionType::from_elements(self.db, [bindings_ty, ty])
+                    }
+                    SymbolLookupResult::Unbound => ty,
+                },
```

Comment:
It feels like we might be reinventing `replace_unbound_with` here? Maybe we should be first building a `SymbolLookupResult` from `may_be_unbound` and `bindings_ty`, and then using `replace_unbound_with` with the results of `self.lookup_name`, and then this code might get a bit clearer / more ergonomic? Not sure.

Reply:
Hm, not quite. Since we need to return a type here, and `replace_unbound_with` returns a `SymbolLookupResult`. I'll look into other way of improving this code.